### ISSUE TITLE
Move to varlink if/when APIs are available

### ISSUE THREAD

**Issue by cooperlees** [cooperlees/monitord#37]
There is world that the dbus calls we use are possibly moving to be exposed via varlink on PID 1 itself. If that happens lets use it as there are less dependencies and cheaper lookups etc. etc.

- The first few versions of monitord will support both and config will drive which is used
- Eventually we'll remove dbus support once a version of systemd with varlink is mainstream

Experiment using async varlink client:
- https://github.com/zeenix/zlink

**Comment by cooperlees** [comment]
Looking today at systemd APIs, still seems we need a lot more implemented before monitord can go anywhere near varlink
- There is also still no async/tokio client

```
[root@b89a51f4c266 systemd]# find . | grep 'io\.'
./userdb/io.systemd.DynamicUser
./userdb/io.systemd.Multiplexer
./userdb/io.systemd.NameServiceSwitch
./userdb/io.systemd.DropIn
./userdb/io.systemd.Home
./io.systemd.ManagedOOM
./io.systemd.Credentials
./journal/io.systemd.journal
./oom/io.systemd.ManagedOOM
./io.systemd.BootControl
./io.systemd.sysext
./netif/io.systemd.Network
./resolve/io.systemd.Resolve
./resolve/io.systemd.Resolve.Monitor
./io.systemd.Hostname
```
- This is with a pretty new systemd too (257.3-7.fc43)

Going to park it for awhile.

**Comment by cooperlees** [comment]
https://github.com/systemd/systemd/issues/36755 - Posted asking for more documentation and a way to see when the APIs we need are available etc. etc.

**Comment by cooperlees** [comment]
https://github.com/systemd/systemd/issues/36877 - Asking for new varlink API to move off parsing the private networkd files we parse ...

**Comment by cooperlees** [comment]
## Remaining varlink work (follow-up to #199–#202)

### Left in monitord — small (same pattern as #200)

_Validated against Fedora Rawhide (systemd 262~rc3) with the varlink-vs-D-Bus parity test, rewritten in Python and runnable locally in #203._

- [x] `UnitsByLoadStateTotal` (v261+) for loaded/masked/not-found instead of per-unit parsing — #204
- [x] `StatusErrno` per-service metric (v261+, feeds full `ServiceStats`) — #205
- [x] `SystemState` + `Version` metrics (or `Manager.Describe`) to kill two D-Bus collectors — #206 (via `Manager.Describe`: `io.systemd.Metrics` has no name filter, so reading two values off the report socket would mean streaming every unit's metrics, ~11ms vs one 3KB reply)
- [ ] Boot/shutdown timestamp family — **split out to #210, to be collected via varlink only.** Two corrections to this entry: monitord collects none of it today (it is genuinely new data, not a parity gap), but it is *not* unavailable over D-Bus — `org.freedesktop.systemd1.Manager` exposes 52 timestamp properties against varlink's 47 families, and our generated proxy already declares all of them unused. Building it varlink-only is a deliberate choice, not a constraint. The two APIs disagree on unset values (varlink emits only what is set, 12 of 47 in the test container; D-Bus returns all as 0), and varlink-only keys will need an explicit parity-test exclusion. Scope and config placement still open — see #210.
- [ ] networkd `Describe` bit-rate fields (v261+) — **blocked: systemd reports nonsense.** With `SpeedMeter=yes` on 262~rc3, `BitRates` comes back as ~1.8e18 bits/s and *identical on all 12 interfaces*, including `lo` and an idle dummy — `networkctl status` renders it `1EBps/1EBps`, so this is networkd's own value, not a varlink or monitord parsing problem. The value also exceeds the `int` the IDL declares for it and grows between samples (looks timestamp-derived), while `/sys/class/net/*/statistics/*` are readable and tiny (rx 1280 bytes). With the default `SpeedMeter=no` the field is absent entirely. So there is nothing worth collecting either way until this is fixed upstream or shown to work on bare metal.
- [x] Time the timer backfill into inner `collection_timings` — #209. Duration lands in `per_unit_loop_ms` and each resolved timer in `timer_dbus_fetches`, on success or failure, so a backfill that dies on `dbus_timeout` is visible rather than silently uncounted. `state_dbus_fetches` is now the only counter that stays 0 on the varlink path.

### Left in monitord — the `Unit.List` adoption project (v258+, fuller detail v259+)

One client (`/run/systemd/io.systemd.Manager`, `io.systemd.Unit.List` streaming + `io.systemd.Manager.Describe`) unlocks all of these:

- [x] Full `ServiceStats` — `runtime.CGroup` (CPU/Memory/Tasks/IO when accounting is on) + `io.systemd.CGroup` metrics + `StatusErrno`; `Type`/`RestartUSec`/`WatchdogUSec`/`TimeoutCleanUSec` from `context.Service`; `MainPID`/`StatusErrno` from `runtime.Service`; process count via `runtime.CGroup.Path` + `cgroup.procs`. Deletes the interim D-Bus oneshot lookup. — #212
- [x] Timers fully varlink-native — `context.Timer` (`AccuracyUSec`, `FixedRandomDelay`, `Persistent`, `RandomizedDelayUSec`, `RemainAfterElapse`, `Unit`) + `runtime.Timer` (`NextElapse*`, `LastTriggerUSec`, `Result`); triggered-service state-change derivable from its own entry. Deletes the D-Bus backfill on host + containers. — #213
- [x] `boot_blame` from per-unit `ActiveTimestamp` enter/exit + `InactiveExitTimestamp` metrics (or `Unit.List` timestamps) — no per-unit calls. — #215 (metrics stream) + #216 (D-Bus fallback for <v261)
- [x] `verify` enumeration from `Unit.List` IDs / metric objects instead of D-Bus `ListUnits`.
- [ ] Machine enumeration via `io.systemd.Machine.List` (`/run/systemd/machine/io.systemd.Machine`, v257+, includes leader pidref) — **blocked by #211.** Per-machine collection cannot use varlink at all: systemd rejects varlink peers from another PID namespace, so every container varlink call through `/proc/<leader>/root/...` is reset and falls back to D-Bus. Enumerating machines over varlink is moot while the collection underneath has to be D-Bus.
  - [x] nspawn fixture investigated — it is what found #211. Booting a machine inside the CI container works (`systemd-container` package + `PrivateUsers=no`), and monitord then collects 38 machine keys from it, all over D-Bus. Landing the fixture in the integration test is still worth doing to lock in the container D-Bus paths from #201/#202/#206.

Out of scope for this list: `dbus_stats` (measures the bus itself), `unit_files` + `pid1` (already D-Bus-free via fs/procfs).

### Needed from systemd (all still open, none requested yet)

- [ ] Emit `ActiveExitTimestamp`/`InactiveEnterTimestamp` in `UnitRuntime` — declared in IDL as `?Timestamp` but absent from all 196 live replies on 262~rc3 (needed for full `ServiceStats` timestamp parity).
- [ ] `io.systemd.Job.List` errors (`NoSuchJob`) on empty queue instead of returning zero replies (workaround exists: `JobsQueued`/`NJobs`).
- [ ] Per-cgroup process count in `CGroup` runtime (workaround exists: `cgroup.procs` under `runtime.CGroup.Path`).
- [ ] Verify on a real host: `io.systemd.CGroup` `List` returned EACCES-as-root in the container while `Basic` worked (likely container artifact).
- [ ] Verify `LastTriggerUSec`/`NRestarts`/`WatchdogUSec`/`IOReadBytes` emission on units where they are nonzero (all null/absent in the test container).

### Version matrix (verified against v260/v261 sources)

- Metrics/report framework (`/run/systemd/report/*`): v260+ base families; v261+ adds `JobsQueued`, `UnitsTotal`, `StateChangeTimestamp`, `ActiveTimestamp`, `InactiveExitTimestamp`, `StatusErrno`, `SystemState`, `UnitsByLoadStateTotal`, `ReloadCount`
- PID 1 socket (`/run/systemd/io.systemd.Manager`): v258+ baseline, v259+ extended, v261+ `Job`/`ListUnitsByNames`/`StartTransient`
- networkd `Describe`: v257+ (bit rates v261+); machined varlink: v257+

**Comment by cooperlees** [comment]
## Road to "zero D-Bus": what claiming it would take (post-#217/#218)

Follow-up to the checklist above, now that verify enumeration (#217) and the
per-section varlink toggles (#218) are merged. This spells out what stands
between here and claiming monitord can run with no D-Bus dependency.

**The claim, precisely.** "Monitord can collect its full stat set with zero
D-Bus traffic" — a runtime property, not "contains zero D-Bus code."
Fallbacks stay until the deprecation plan below fires: old-systemd and
embedded support is a goal, and the varlink floors range v257–v261, so the
claim would read "on a new-enough systemd, with the right config, the bus is
never touched." Mandatory carve-out: `[dbus]` measures the bus itself, so it
must be disabled for the claim to hold (or scoped out of it the way
`dbus_stats` already is from the parity work).

### Host ledger (systemd v261+ — the max floor, see versions)

Done — varlink-native with D-Bus fallback retained:
- `units` + `timers` + `services`: metrics stream + `Unit.List` (#212 full
  `ServiceStats`, host oneshot D-Bus lookup deleted; #213 timers fully
  varlink-native, host backfill deleted)
- `networkd`: `Network.Describe` (v257+) — touches no files on disk on the
  varlink path
- `system-state` + `version`: `Manager.Describe` (v258+)
- `boot_blame`: metrics `ActiveTimestamp`/`InactiveExitTimestamp` (v261+),
  D-Bus fallback below (#216)
- `verify`: varlink enumeration (#217); `systemd-analyze` itself is a
  subprocess either way, on both paths
- `pid1` (procfs) and `unit_files` (fs walk) never touched D-Bus

Remaining on host:
1. **Eager D-Bus connection (`lib.rs`).** `stat_collector` builds the system
   bus connection up front and `?`s on failure — so today monitord cannot
   even start without a bus, regardless of which collectors are enabled.
   Prerequisite for any proof run: connect lazily (only when a D-Bus path is
   actually taken) or tolerate connect failure when nothing enabled needs
   the bus.
2. **Fallbacks still use D-Bus by design** (e.g. networkd's `list_links`).
   Fine for old systemd; the "no D-Bus" config just has to keep every
   collector on varlink. Enforcement comes from strict mode below, not from
   deleting fallback code.

### Containers: #211 is the wall

#211 (open): systemd rejects varlink peers from another PID namespace, so
every container varlink call through `/proc/<leader>/root/...` is reset and
falls back to D-Bus. Consequence: **all** container collection — units,
networkd, system-state, version — is D-Bus plus D-Bus backfills (timer
backfill, oneshot `Type` override) by necessity, and machine enumeration
itself is D-Bus `list_machines`. Varlink `io.systemd.Machine.List` exists
(v257+, includes the leader pidref) but adopting it is moot while everything
underneath has to be D-Bus anyway.

Until upstream allows host→container varlink (or documents another
mechanism), the zero-D-Bus claim can only cover `[machines] enabled = false`.
This is the single biggest item and it is entirely upstream.

### Upstream asks

Resolved upstream:
- [systemd#36877](https://github.com/systemd/systemd/issues/36877) — closed;
  `Network.Describe` covers us and the varlink path parses no files.

Still filed/tracked:
- [systemd#36755](https://github.com/systemd/systemd/issues/36755) — varlink
  docs + API availability signals

Still to request (from the earlier list, all unverified on current Rawhide):
- `ActiveExitTimestamp`/`InactiveEnterTimestamp` emission in `UnitRuntime`
  (declared `?Timestamp`, absent from all live replies)
- `io.systemd.Job.List` errors on empty queue instead of zero replies
- Per-cgroup process count in `CGroup` runtime
- `io.systemd.CGroup` `List` EACCES-as-root vs `Basic` (likely container
  artifact — needs a bare-metal check)
- `LastTriggerUSec`/`NRestarts`/`WatchdogUSec`/`IOReadBytes` emission where
  nonzero (all null/absent in the test container)
- The #211 need itself: no systemd issue filed yet — needs a minimal
  reproducer and a concrete proposal (allow cross-namespace peers? a proxy
  mechanism?) before filing

### Verification plan (how to prove the claim in CI)

1. **Bus-blocked run**: point the bus address at a nonexistent socket, enable
   varlink globally + all host collectors, disable `[dbus]` + `[machines]`;
   assert exit 0 and a complete key set. Blocked on host item 1 (lazy
   connection). This is the proof run the claim rests on.
2. **Strict mode**: add a `varlink_strict` flag that errors instead of
   falling back, and run CI in that mode. Per-section toggles (#218) already
   allow flipping one collector at a time; strict mode makes any fallback a
   loud failure instead of a silent one.
3. **Zero-fetch assertion**: extend the parity test to assert every
   `*_dbus_fetches` counter is 0 on the varlink run (the no-fallback-log
   assertion already exists; the counters make it structural).
4. **Version floor**: v261+ for the full claim (`StatusErrno`, the timestamp
   families boot blame reads, `JobsQueued`, exact totals). Anything older
   demonstrably splits — e.g. Ubuntu's systemd 259 has `Describe` but no
   metrics socket, so system-state/version go varlink while units/verify/boot
   stay D-Bus.

### Deprecation plan

Once CI proves 100% varlink-only via strict mode, publicize the plan to
deprecate D-Bus support once most mainstream distributions ship a new-enough
systemd (v261+ for the full claim). Until that bar is met, D-Bus fallbacks
stay and stay tested — the parity test keeps both paths honest in the
meantime.

### Open decisions

- **Claim scope**: host-only + `[machines]` off until #211 moves?
  (Recommended: yes — state it explicitly rather than letting containers
  silently disqualify the claim.)
- **Compile time**: keep the zbus dependency unconditionally (`dbus_stats` +
  fallbacks need it) — i.e. the claim stays runtime-only. Agreed?
- **#210** (boot/shutdown timestamps, planned varlink-only new data): no
  D-Bus involved, so it neither helps nor hurts the claim — just don't add a
  D-Bus path for it later.

### Status note

The `verify` enumeration box in the checklist above is done (#217 merged);
only the #211-gated machine work remains open on the monitord side.

**Comment by cooperlees** [comment]
## systemd#43807 filed for #211; triage of the remaining upstream asks

### Machine stats / #211 — now filed

[systemd/systemd#43807](https://github.com/systemd/systemd/issues/43807) — "Allow Varlink peers from a different (ancestor) PID namespace, or make the peer-credential requirement configurable".

This closes the last open item from the previous comment's "Still to request" list ("the #211 need itself: no systemd issue filed yet — needs a minimal reproducer and a concrete proposal before filing"). Both prerequisites are met:

- **Mechanism confirmed**, not inferred. `getpeercred()` returns `-ENODATA` because `SO_PEERCRED` reports `pid == 0` when the peer's PID is not translatable into the server's PID namespace; `sd_varlink_server_add_connection_pair()` then refuses pre-handshake. Only servers setting `ROOT_ONLY`/`MYSELF_ONLY`/`ACCOUNT_UID` do the lookup — PID 1 sets `ACCOUNT_UID` for system managers, networkd unconditionally. Full detail in #211.
- **Minimal reproducer**: two Docker containers, no nspawn and no machine image, because Docker already puts the container's systemd in a descendant PID namespace. Also in #211.
- **Concrete proposal**: four options, framed as opt-in so the current default need not change — relax only where the PID is unused (the policy reads only `ucred.uid`, never `ucred.pid`), a new `SD_VARLINK_SERVER_*` flag, a `system.conf`/`networkd.conf` knob, or a documented statement that it is intentional. The last is explicitly called out as an equally useful outcome, since it lets us stop attempting per-container varlink with a definitive reason.

**Status of the blocked checklist item is unchanged**: machine enumeration via `io.systemd.Machine.List` stays blocked, and the "host-only + `[machines]` off" scope for the zero-D-Bus claim still looks right. Even a favourable upstream ruling lands no earlier than a future systemd release, so #211's option 1 (stop the per-cycle varlink attempts, or probe once) should proceed independently rather than waiting on #43807.

Incidental cross-check: [systemd#43175](https://github.com/systemd/systemd/issues/43175) (someone else's job-subscription RFE) independently documents that PID 1 sets `SD_VARLINK_SERVER_ACCOUNT_UID` and enforces `VARLINK_DEFAULT_CONNECTIONS_PER_UID_MAX` = 128. That corroborates the flag analysis in #43807, and the 128-per-UID ceiling is worth remembering if we ever fan out varlink connections per machine.

### Searched upstream for the other asks — results

Searched systemd/systemd issues **and** PRs, open and closed, for each remaining item. Caveat: GitHub search is keyword-based, so no-hits is weak evidence rather than proof.

| Ask | Upstream coverage | Workaround? | Recommendation |
|---|---|---|---|
| `ActiveExitTimestamp`/`InactiveEnterTimestamp` absent from `UnitRuntime` | **none found** | no | **File.** Only one of these that is a real parity gap with no way around it |
| `io.systemd.Job.List` returns zero replies on empty queue instead of erroring | none for this semantic; #43175 is an adjacent but different Job RFE | yes — `JobsQueued`/`NJobs` | Don't file standalone |
| Per-cgroup process count in `CGroup` runtime | **none found** | yes — `cgroup.procs` under `runtime.CGroup.Path` | Optional, low priority |
| `io.systemd.CGroup` `List` EACCES-as-root while `Basic` works | **none found** | n/a | **Verify on bare metal first** — likely a container artifact; filing a container-only symptom wastes everyone's time |
| `LastTriggerUSec`/`NRestarts`/`WatchdogUSec`/`IOReadBytes` null where they should be set | none for varlink emission (#29348 touches `NRestarts` semantics, not emission) | n/a | **Verify first** on a host where these are genuinely nonzero — cannot distinguish "not emitted" from "correctly zero in the test container" |

So: **one issue worth filing now** (the `UnitRuntime` timestamp pair), two that are documented workarounds not worth upstream traffic, and two that are unverified observations rather than confirmed bugs. The two "verify first" items both need a bare-metal Rawhide run — worth batching into one session, and if either reproduces off-container they become filings.

The timestamp one is the natural next filing: it is a concrete IDL-versus-behaviour mismatch (declared `?Timestamp`, absent from all 196 live replies on 262~rc3), which is the same shape as #43807 and easy for a maintainer to confirm or dismiss quickly.

### Descoped: networkd bit-rate fields

Dropping the `Describe` bit-rate entry from the varlink work. monitord does not collect bit rates today over **any** transport, so this was never a parity gap — it is new data that happened to surface while auditing `Describe`. Excluding it keeps the zero-D-Bus claim about parity with what we actually ship, instead of entangling it with an upstream networkd value that looked wrong on 262~rc3.

If bit rates are ever wanted, they come back as a new-feature request on their own merits, with the upstream value re-checked at that point. No systemd issue from us in the meantime.

**Comment by cooperlees** [comment]
## "Needed from systemd" verified — four of five were not systemd bugs

Follow-up to the triage comment above. Every item on the "Needed from systemd" list has now been checked against a live Fedora Rawhide container (systemd `262~rc3-1.fc46`) and against systemd's source, **before** filing anything. Four of the five turned out not to be bugs, so filing them as written would have sent bad reports upstream.

| Ask (as written above) | Observed | Verdict |
|---|---|---|
| `ActiveExitTimestamp`/`InactiveEnterTimestamp` "absent from all 196 live replies" | **Present** — 4 and 8 of 308 units, precisely those that had exited active / entered inactive | **Retracted** |
| `io.systemd.Job.List` errors on empty queue | **Confirmed** — `io.systemd.Job.NoSuchJob` on an idle system | **Real** → filed |
| Per-cgroup process count missing from `CGroup` runtime | `pids.current` provides it directly from cgroupfs | **Retracted** — no upstream ask needed |
| `io.systemd.CGroup` `List` EACCES-as-root | **Not reproducible** — there is no such interface | **Retracted** |
| `LastTriggerUSec`/`NRestarts`/`WatchdogUSec`/`IOReadBytes` null where nonzero | All **emitted** once genuinely set | **Retracted** |

### Why each retraction happened

- **The timestamp pair.** Both are declared `?Timestamp` in the IDL, and varlink omits optional fields when unset. Most units boot, go active and never exit, so the fields are correctly absent. The original "absent from all 196 replies" was a sampling artifact of a container where nothing had cycled.
- **`NRestarts` / `IOReadBytes`.** Same class of error, confirmed by forcing the conditions. A transient unit with `Restart=always` reached `NRestarts=5`, and `Unit.List` emitted `"NRestarts":5`. A unit with `IOAccounting=yes` caused systemd to add `io` to `system.slice/cgroup.subtree_control`, `io.stat` appeared, and `IOReadBytes` was emitted. Both were previously zero because nothing had restarted and the `io` controller was not enabled — not because systemd withholds them.
- **`io.systemd.CGroup` EACCES.** There is no `io.systemd.CGroup` interface. `/run/systemd/report/io.systemd.CGroup` serves `io.systemd.Metrics`, and it returns `NoSuchMetric`, not EACCES. `src/report/report-cgroup.c` emits only for paths where `cg_path_get_unit()` succeeds and **stops at delegation boundaries** (`cg_is_delegated()`); our test container runs `--cgroupns=host`, so its units sit under `/sys/fs/cgroup/docker/<id>/system.slice/...` behind Docker's delegated cgroup. A container artifact, exactly as this list suspected.
- **Per-cgroup process count.** `pids.current` is a plain file. See the cgroupfs section below.

### Filed upstream

- [systemd#43807](https://github.com/systemd/systemd/issues/43807) — cross-PID-namespace varlink peers (the #211 blocker), filed earlier.
- [systemd#43808](https://github.com/systemd/systemd/issues/43808) — **new**: varlink `List` methods should return an empty stream rather than erroring when there is nothing to list.

The second one groups two symptoms into a single issue because they are the same design question in the same component: `io.systemd.Job.List` raises `NoSuchJob` on an idle system, and `io.systemd.Metrics.List` raises `NoSuchMetric` on a report socket with nothing to report (`CGroup`, `Files`), while `Basic`/`Manager`/`Network` stream normally. In both cases the error is defined for a lookup that named something specific — `NoSuchJob` is documented as "The specified job does not exist", and `Metrics.List` takes no parameters at all — so neither can refer to anything the caller actually asked for.

**With that, "Needed from systemd" is closed apart from systemd#43807 and systemd#43808.**

### The cgroup finding — a better option than any upstream ask

`systemd-report-cgroup` reads **plain cgroup v2 files directly** (`cpu.stat`, `io.stat`, `memory.current`, `memory.max/high/peak`, `pids.current`) via `cg_get_keyed_attribute()` / `cg_get_attribute_as_uint64()`. No accounting API underneath. Confirmed readable with no systemd involvement:

```
cpu.stat        usage_usec 210697, user_usec 151702, system_usec 58995
memory.current  3829760
memory.peak     6471680
pids.current    1
```

Seven `ServiceStats` fields are cgroup-derived and all of them map onto those files. They are currently D-Bus-only — `io.systemd.CGroup` does not exist and the report socket is unusable as shown above — so they are a hard blocker for the zero-D-Bus claim. Reading them ourselves removes that blocker, drops the dependency on accounting flags, and (because container unit cgroups are nested in the host hierarchy) works for machines **without** the #211 PID-namespace wall. We already do exactly this for one field in `count_cgroup_processes()`.

Opened as **#221**.

### Also opened

**#220** — `src/machines.rs:375` logs the container unit-stats fallback without the machine name, unlike its three siblings, and byte-identically to the host message at `src/lib.rs:512`. Given #211 makes every container emit it every cycle, it is the most frequent and least informative line in the log.

### Documented in AGENTS.md

A new `### Known Limitations` section records the container/varlink PID-namespace limitation and points at #211 / systemd#43807, plus three conventions: how to reference upstream issues, to check for `?Type` optionality before reporting a field as missing (the mistake that produced four bad entries on this list), and to prefer kernel-exposed data over IPC.
